### ISSUE TITLE
Config hash can't contain array of hashes

### ISSUE THREAD

**Issue by mathiasaerts** [elastic/puppet-elasticsearch#618]
I'm installing an ES plugin via puppet that uses a couple of parameters in the `elasticsearch.yml` file. Some of these parameters are supposed to be made up out of an array of hashes, like this:
```ruby
data => [ { key       => "some value 0",
            other_key => "some other value 0" },
          { key       => "some value 1",
            other_key => "some other value 1" } ]
```
This should result in the following yaml snippet:
```ruby
  data:
    - key: some value 0
      other_key: some other value 0
    - key: some value 1
      other_key: some other value 1
```

However, when I added these parameters to the `config` hash that is passed to the `elasticsearch::instance` define, puppet threw me the following error: "Could not evaluate: can't convert Hash into String". 

After checking the `elasticsearch.yml.erb` template I noticed that the `recursive_hash_to_yml_string` method doesn't check if an array value is a hash, so I changed the method to the following:

```ruby
  # Function to make a structured and sorted yaml representation out of a hash
  def recursive_hash_to_yml_string(hash, depth=0)
    spacer = ""
    depth.times { spacer += "  "}
    hash.keys.sort.each do |sorted_key|
      @yml_string += spacer + sorted_key + ": "
      if hash[sorted_key].is_a?(::Array)

         keyspacer = "  "
         # I can't see why to do this, replaced by keyspacer = "  "
         #sorted_key.length.times { keyspacer += " " }

         @yml_string += "\n"
         hash[sorted_key].each do |item|

            # Is the current item a hash?
            if item.is_a?(::Hash)
               @yml_string += spacer + keyspacer + "- \n"
               recursive_hash_to_yml_string(item, depth+2)
            else
               @yml_string += spacer + keyspacer + "- " + item +"\n"
            end

         end
      elsif hash[sorted_key].is_a?(::Hash)
        @yml_string += "\n"
        recursive_hash_to_yml_string(hash[sorted_key], depth+1)
      else
        @yml_string += "#{hash[sorted_key].to_s}\n"
      end
    end
  end
```

Given the array above, it will now generate the following yaml code:
```ruby
  data:
    -
      key: some value 0
      other_key: some other value 0
    -
      key: some value 1
      other_key: some other value 1
```
This includes an extra newline after the `-` but it seems ES as well as the plugin are perfectly happy with this!

**Comment by tylerjl** [comment]
I'd definitely prefer to have the erb template spit out the yaml without the newlines if possible. If you can get a PR together that has your change plus some tests for hashes in config arrays and everything is green, that's good enough for me. 👍  Thanks for taking the time to fix it!

**Comment by mathiasaerts** [comment]
Fixing the newlines would take some more time. The problem is that only the first item of the hash should use different spacing. The hash inside the array is passed to the function again recursively, and there is no way of knowing if it was originally inside an array or not (unless we add an extra parameter to the method).

However, I've looked up the Yaml spec and found the example below:
http://www.yaml.org/spec/1.2/spec.html#id2759963 (Example 2.4.  Sequence of Mappings)

Here you can see the same output with newlines, so I think Yaml definitely supports this.

I'll see if I can find the time to create a pull request and write some tests.

**Comment by tylerjl** [comment]
@mathiasaerts I don't know if you're still looking at this, but I've been considering replacing the [template yaml function](https://github.com/elastic/puppet-elasticsearch/blob/995c1a9e92ead8cb33e5031e0a67cde8a5c49eb1/templates/etc/elasticsearch/elasticsearch.yml.erb#L4) just entirely with puppet's native `to_yaml`, since other steps in the template already handle breaking up the dots in the config hash and merging duplicate keys. It would probably help to avoid edge cases like this with a custom to_yaml function.

#630 has tests for the template now to verify it functions correctly, so this may be easier to do.

**Comment by 1n** [comment]
I'm also interested in resolving this issue.

I've spent some time to replace template yaml function with native to_yaml.
https://github.com/tylerjl/puppet-elasticsearch/compare/feature/merge_qualified_configs...1n:bug/array_of_hashes?expand=1

It works for puppet 3.8.0 and tests are passing.
But puppet > 4.0.0 has different YAML implementation.
While both versions render a valid YAML.

```shell

$ export PUPPET_VERSION='~> 4.3.0'

$ make test-rspec
...
Failures:

  1) elasticsearch.yml.erb should render normal hashes
     Failure/Error:
       expect(harness.run).to eq(%q{
         discovery:
           zen:
             ping:
               unicast:
                 hosts:
                   - host1
                   - host2
         node:
           name: test
     
       expected: "### MANAGED BY PUPPET ###\n---\ndiscovery:\n  zen:\n    ping:\n      unicast:\n        hosts:\n          - host1\n          - host2\nnode:\n  name: test\npath:\n  data: /mnt/test\n"
            got: "### MANAGED BY PUPPET ###\n---\ndiscovery:\n  zen:\n    ping:\n      unicast:\n        hosts:\n        - host1\n        - host2\nnode:\n  name: test\npath:\n  data: /mnt/test\n"
     
       (compared using ==)
     
       Diff:
       @@ -5,8 +5,8 @@
            ping:
              unicast:
                hosts:
       -          - host1
       -          - host2
       +        - host1
       +        - host2
        node:
          name: test
        path:
       
     # ./spec/templates/001_elasticsearch.yml.erb_spec.rb:32:in `block (2 levels) in <top (required)>'

  2) elasticsearch.yml.erb should concatenate arrays
     Failure/Error:
       expect(harness.run).to eq(%q{
         data:
           path:
             - /mnt/sda1
             - /mnt/sdb1
         }.config)
     
       expected: "### MANAGED BY PUPPET ###\n---\ndata:\n  path:\n    - /mnt/sda1\n    - /mnt/sdb1\n"
            got: "### MANAGED BY PUPPET ###\n---\ndata:\n  path:\n  - /mnt/sda1\n  - /mnt/sdb1\n"
     
       (compared using ==)
     
       Diff:
       @@ -2,6 +2,6 @@
        ---
        data:
          path:
       -    - /mnt/sda1
       -    - /mnt/sdb1
       +  - /mnt/sda1
       +  - /mnt/sdb1
       
     # ./spec/templates/001_elasticsearch.yml.erb_spec.rb:70:in `block (2 levels) in <top (required)>'

  3) elasticsearch.yml.erb should qualify conflicting hash keys
     Failure/Error:
       expect(harness.run).to eq(%q{
         shield:
           http:
             ssl: "true"
             ssl.client:
               auth: optional
         }.config)
     
       expected: "### MANAGED BY PUPPET ###\n---\nshield:\n  http:\n    ssl: \"true\"\n    ssl.client:\n      auth: optional\n"
            got: "### MANAGED BY PUPPET ###\n---\nshield:\n  http:\n    ssl: 'true'\n    ssl.client:\n      auth: optional\n"
     
       (compared using ==)
     
       Diff:
       @@ -2,7 +2,7 @@
        ---
        shield:
          http:
       -    ssl: "true"
       +    ssl: 'true'
            ssl.client:
              auth: optional
       
     # ./spec/templates/001_elasticsearch.yml.erb_spec.rb:86:in `block (2 levels) in <top (required)>'

  4) elasticsearch.yml.erb should render correct array of hashes
     Failure/Error:
       expect(harness.run).to eq(%q{
         data:
           - key: "some value 0"
             other_key: "some other value 0"
           - key: "some value 1"
             other_key: "some other value 1"
         discovery:
           zen:
             ping:
               unicast:
     
       expected: "### MANAGED BY PUPPET ###\n---\ndata:\n  - key: \"some value 0\"\n    other_key: \"some other value 0\"\n  - key: \"some value 1\"\n    other_key: \"some other value 1\"\ndiscovery:\n  zen:\n    ping:\n      unicast:\n        hosts:\n          - host1\n          - host2\nnode:\n  name: test\npath:\n  data: /mnt/test\n"
            got: "### MANAGED BY PUPPET ###\n---\ndata:\n- key: some value 0\n  other_key: some other value 0\n- key: some value 1\n  other_key: some other value 1\ndiscovery:\n  zen:\n    ping:\n      unicast:\n        hosts:\n        - host1\n        - host2\nnode:\n  name: test\npath:\n  data: /mnt/test\n"
     
       (compared using ==)
     
       Diff:
       
       @@ -1,17 +1,17 @@
        ### MANAGED BY PUPPET ###
        ---
        data:
       -  - key: "some value 0"
       -    other_key: "some other value 0"
       -  - key: "some value 1"
       -    other_key: "some other value 1"
       +- key: some value 0
       +  other_key: some other value 0
       +- key: some value 1
       +  other_key: some other value 1
        discovery:
          zen:
            ping:
              unicast:
                hosts:
       -          - host1
       -          - host2
       +        - host1
       +        - host2
        node:
          name: test
        path:
       
     # ./spec/templates/001_elasticsearch.yml.erb_spec.rb:112:in `block (2 levels) in <top (required)>'

Finished in 0.07991 seconds (files took 1.49 seconds to load)
5 examples, 4 failures
```
What is the desired behavior?
Should it be identical for both puppet versions? Or how could this difference in YAML implementation be handled in rspec tests?

**Comment by tylerjl** [comment]
@1n thanks for doing the work on that, the differences between 3/4 are especially annoying.

Honestly, as long as it's valid yaml, that's really all that matters, really. My gut feeling on this is that we should probably try and handle subtle differences in the erb template, and let the rspec tests stay the same to confirm our yaml rendering is taking care of whatever oddities Puppet comes up with.

Another option may be to re-parse the yaml in the rspec tests and compare the resulting hashes, which doesn't explicitly test the exact contents of the yaml config file but should be functionally what Elasticsearch sees anyway. You're doing good work with the tesitng and implementation, I'm open to whichever works better.

**Comment by mathiasaerts** [comment]
> @tylerjl I've been considering replacing the template yaml function just entirely with puppet's native to_yaml

I did try this at some point before I changed the function in the erb template. However, the output wasn't what I expected, adding quotes around values if I remember correctly. Elasticsearch wasn't very happy with the generated yaml, so I decided to revert those changes. Maybe it was only a small issue, I didn't really look into it at that point.

I also think using `to_yaml` would be better, and it seems like @1n has done a great job already.
